dipy/reconst/rumba.py

```diff
@@ -0,0 +1,815 @@
+    '''
+    Generate deconvolution kernel
+
+    Compute kernel mapping volume fractions of white matter fiber populations (oriented along each
+    vertex of a sphere) and cerebrospinal fluid (CSF) to a diffusion weighted signal.
+
+    Parameters
+    ----------
+    gtab : GradientTable
+    sphere : Sphere
+        sphere with which to sample discrete fiber orientations in order to construct kernel
+    lambda1 : float, optional
+        white matter diffusivity parallel to fiber axis (first DTI eigenvalue). Default: 1.7e-3
+    lambda2 : float
+        white matter diffusivity perpendicular to fiber axis (second/third DTI eigenvalues are
+        assumed equal). Default: 0.2e-3
+    lambda_csf : float
+        Mean diffusivity for CSF. If `None`, then CSF kernel column set to all zeroes.
+        Default: 3.0e-3
+
+    Returns
+    -------
+    kernel : 2d ndarray (N, M)
+        computed kernel; can be multiplied with a vector consisting of volume fractions for each of
+        M-1 fiber populations and the fraction of CSF to produce a diffusion weighted signal.
+    '''
+
+    # Coordinates of sphere vertices
+    _, theta, phi = cart2sphere(
+        sphere.x,
+        sphere.y,
+        sphere.z
+    )
+
+    n_grad = len(gtab.gradients)  # number of gradient directions
+    n_wm_comp = len(theta)  # number of fiber populations
+    n_comp = n_wm_comp + 1  # plus compartment for CSF
+
+    kernel = np.zeros((n_grad, n_comp))
+
+    S0 = 1  # S0 assumed to be 1
+    fi = 100  # volume fraction assumed to be 100%
+
+    # White matter components
+    for i in range(n_wm_comp):
+        # Convert angles to degrees
+        angles = [theta[i] * 180 / np.pi, phi[i] * 180 / np.pi]
+
+        # Signal generated by WM-fiber for each gradient direction
+        S, _ = multi_tensor(gtab, np.array([[lambda1, lambda2, lambda2]]),
+                            S0, [angles], [fi], None)
+        kernel[:, i] = S
+
+    # CSF component
+    if lambda_csf is None:
+        S_csf = np.zeros((n_grad))
+    else:
+        S_csf, _ = multi_tensor(gtab, np.array([[lambda_csf, lambda_csf, lambda_csf]]),
+                                S0, [[0, 0]], [100], None)
+
+    kernel[:, n_wm_comp] = S_csf
+
+    return kernel
+
+
+def combine_odf_csf(odf, f_csf):
+    '''
+    Combine fODF and CSF volume fraction maps
+
+    Distributes CSF volume fraction at each voxel along each fODF direction
+
+    Parameters
+    ----------
+    odf : 4d ndarray (x, y, z, K)
+        fODF for each voxel
+    f_csf : 3d ndarray (x, y, z)
+        volume fraction of CSF for each voxel
+
+    Returns
+    -------
+    combined : 4d ndarray (x, y, z, K)
+        fODF for each voxel combined with CSF volume fractions
+    '''
+    combined = odf + f_csf[..., None] / odf.shape[3]
+    return combined
+
+
+### GLOBAL FITTING ###
+
+def global_fit(model, data, sphere, mask=None, use_tv=True, verbose=False):
```

Comment:
Instead of `verbose`, replace all your print by `logging.info(msg)`. If the user wants a verbose output, he can set `logger.setLevel(logging.INFO). 

This is the way it is done in dipy.tracking.streamline.py. Elsewhere `logging.getLogger(__name__)` is used. I'm not sure which one is best.

Reply:
As per what's done in segment.bundles, segment.clustering, align.imwarp, align.streamlinear, I kept the verbose parameter and, if verbose, I use the logger initialized by logging.getLogger(__name__) to log the updates at the info level.

Reply:
ok!